book/src/concepts/constructor.md

```diff
@@ -0,0 +1,148 @@
+```
+[Full example](https://github.com/KDAB/cxx-qt/blob/main/examples/qml_features/rust/src/signals.rs)
+
+## Routing arguments of the Constructor
+A C++ constructor is more complex then a Rust struct initialization.
+
+Like a normal function, it takes a list of arguments.
+However, before the actual code inside the constructor is executed, C++ ensures the following:
+* The base class is constructed
+* All members of the struct are initialized (for CXX-Qt this includes the underlying Rust struct).
+
+Both of these steps may need any of the arguments given to the constructor.
+The code that is actually run *inside* the constructor may also need to access some of the arguments.
+
+`cxx_qt::Constructor` allows controlling how the arguments are passed around in the constructor using the `route_arguments` function.
+For this it uses three associated types as well as one generic type:
+* Constructor generic `Arguments` - argument list that the constructor takes.
+* `type BaseArguments` - argument list that is passed to the base class constructor.
+* `type NewArguments` - argument list used to construct the Rust struct using the `new` function.
+* `type InitializeArguments` - argument list that is passed to the code run within the constructor (the `initialize` function).
+
+These types must all be tuple types and can therefore support an arbitrary list of arguments.
+
+The `route_arguments` function takes the Arguments provided to the constructor and distributes them to the three different types in the above list.
+Doing this in Rust allows calling any base class constructor and controlling movement, copying or other operations on the arguments.
+Every constructor defined using `cxx_qt::Constructor` will execute this function once, before anything else in the Constructor happens and then distributes the resultant values to the appropriate steps of the QObject construction.
+
+As actual construction takes place in C++, all types used must be compatible with CXX as they must be able to pass the FFI barrier.
+
+### Default constructor
+To implement a default constructor in CXX-Qt (the constructor used by QML) a good default would be to just leave everything as `()`.
+The `route_arguments` implementation would then look like this:
+```rust,ignore
+impl cxx_qt::Constructor<()> for qobject::T {
+    type NewArguments = ();
+    type InitializeArguments = ();
+    type BaseArguments = ();
+
+    fn route_arguments(_arguments: ()) -> (
+        Self::NewArguments,
+        Self::BaseArguments,
+        Self::InitializeArguments,
+    ) {
+        // Nothing to route
+        ((),(),())
+    }
+}
+```
+
+## Constructing the Rust struct with arguments
+If the construction of the inner Rust struct requires some arguments, the `new` function of the Constructor can be given arguments by using the `NewArguments` associated type.
+
+Example:
+```rust,ignore
+impl cxx_qt::Constructor<(i32,)> for qobject::T {
```

Comment:
do we not have an examples where these can be pulled from ? As we've been trying to remove blocks of untested/built code from the book / docs

Reply:
Currently not, as qml_features uses QML, which always default-constructs anyway.
I can add a dummy constructor to properties to make sure it compiles though.

Reply:
:thinking: just write a comment next to it saying the dummy one has no purpose other than for documentation so folks don't think it's required :-) 

Was just trying to ensure we don't add more code blocks that could be tested.